### ISSUE TITLE
[WIP] Sglang-Omni refractor

### ISSUE THREAD

**Issue by JingwenGu0829** [sgl-project/sglang-omni#227]
# SGLang-Omni refractor design, still working in progress. Implementing concrete code to test optimal design.

Everything follows mainly from https://github.com/sgl-project/sglang/issues/16546. Trying to align the initial RFC with the actual engineering obstacles we encountered.
---

### Request Path & Component Interaction

```mermaid
flowchart TB
    subgraph "Request Path"
        HTTP["HTTP API"] --> Client --> Coordinator
    end

    subgraph "Pipeline Layer"
        Coordinator -->|SubmitMessage| StageA["Stage"]
        StageA -->|"relay + DataReadyMessage"| StageB["Stage"]
        StageB -->|"relay + DataReadyMessage"| StageC["Stage"]
        StageC -->|CompleteMessage| Coordinator
    end

    subgraph "Inside Each Stage"
        direction TB
        CP["Control Plane\n(ZMQ)"] -->|recv| Stage
        Stage -->|"inbox.put()"| Scheduler
        Scheduler -->|"outbox.put()"| Stage
        Stage -->|write| DP["Data Plane\n(Relay: SHM/NCCL/NixL)"]
    end

    subgraph "Scheduler Types"
        direction LR
        OmniSched["OmniScheduler\n(inherits SGLang Scheduler)\nAR stages: thinker, talker, fish TTS"]
        SimpleSched["SimpleScheduler\ninbox→fn()→outbox\npreprocessing, encoders, decode"]
        C2WSched["Code2WavScheduler\nstreaming vocoder\naccumulate chunks → decode"]
    end

    subgraph "Model Runner Layer"
        direction TB
        MR["ModelRunner (base)\nexecute pipeline:\nForwardBatch → prepare_forward → forward → post_forward → sample"]
        MR --> ThinkerMR["ThinkerModelRunner\nprepare: inject multimodal embeds"]
        MR --> FeedbackMR["FeedbackARModelRunner\nprepare: write_buffers_fn()\npost: extract_output_fn()"]
    end

    subgraph "Model-Specific Callbacks"
        direction LR
        FeedbackMR --> QwenCB["Qwen3 callbacks.py\nwrite_talker_buffers\nextract_talker_output\ntalker_prefill_forward"]
        FeedbackMR --> FishCB["Fish callbacks.py\nwrite_fish_buffers\nextract_fish_output\nfish_prefill_forward"]
        FeedbackMR --> FutureCB["Future model\ncallbacks.py\n(3 functions)"]
    end

    OmniSched -->|"run_batch()"| MR
```

### Qwen3-Omni Speech Pipeline (8 stages)

```mermaid
flowchart LR
    Pre["preprocessing\nSimpleScheduler"] --> ImgEnc["image_encoder\nSimpleScheduler"]
    Pre --> AudEnc["audio_encoder\nSimpleScheduler"]
    Pre --> Agg["aggregate\nSimpleScheduler"]
    ImgEnc --> Agg
    AudEnc --> Agg
    Agg --> Thinker["thinker\nOmniScheduler\n+ ThinkerModelRunner"]
    Thinker -->|payload| Decode["decode\nSimpleScheduler"]
    Thinker -->|"stream hidden states"| Talker["talker\nOmniScheduler\n+ FeedbackARModelRunner"]
    Thinker -->|payload| Talker
    Talker -->|"stream codes"| C2W["code2wav\nCode2WavScheduler"]
```

### Fish Audio S2-Pro Pipeline (3 stages)

```mermaid
flowchart LR
    FPre["preprocessing\nSimpleScheduler"] --> TTS["tts_engine\nOmniScheduler\n+ FeedbackARModelRunner"]
    TTS --> Voc["vocoder\nSimpleScheduler"]
```

### Model Directory Convention

```mermaid
flowchart LR
    Config["config.py\npipeline config"] --- Stages["stages.py\nfactory per stage"]
    Stages --- Routing["routing.py\nnext_stage functions"]
    Routing --- RB["request_builders.py\ninter-stage data transform"]
    RB --- PT["payload_types.py\nmodel-specific state"]
    PT --- CB["callbacks.py\nFeedbackARModelRunner hooks"]
    CB --- Comp["components/\ntorch modules\npreprocessor\nencoders"]
```

---

## 1. System Overview

```
HTTP API → Client → Coordinator → Stage → [Scheduler → ModelRunner → forward]
```

```
sglang_omni/
├── pipeline/           # Inter-stage orchestration (model-agnostic)
├── scheduling/         # Scheduling loops (OmniScheduler, SimpleScheduler)
├── model_runner/       # Model runner base + shared FeedbackARModelRunner
├── models/             # Model definitions + pipeline configs
├── config/             # Pipeline config schema + compiler
├── relay/              # Data transfer backends (SHM, NCCL, NixL)
├── serve/              # HTTP server, OpenAI API
├── client/             # Client library
└── proto/              # Message types
```

### 2.1 Coordinator

Global request router. Tracks request lifecycle across stages.

- Routes new requests to entry stage
- Collects completions from terminal stages
- Merges results when multiple terminal stages exist (e.g., decode + code2wav)
- Broadcasts abort to all stages

### 2.2 Stage

IO shell. Every stage has one scheduler (no branching). Handles all inter-stage communication.

```python
class Stage:
    def __init__(self, name, control_plane, relay, get_next,
                 input_handler, scheduler, stream_targets, same_gpu_targets):
        self.scheduler = scheduler  # always present
```

Responsibilities:
- **Control plane** (ZMQ): receive SubmitMessage, DataReadyMessage, AbortMessage
- **Data plane** (Relay): read/write tensors between stages via SHM/NCCL/NixL
- **Input aggregation**: wait for multiple upstream stages before dispatching (AggregatedInput)
- **Stream routing**: receive/send streaming chunks (hidden states, codec codes)
- **Dispatch**: push all messages into `scheduler.inbox`, drain `scheduler.outbox`

One code path. Stage never checks scheduler type.

### 2.3 Inter-Stage Communication

```
Stage A:  relay.write(tensors) → control_plane.send(DataReadyMessage)
Stage B:  control_plane.recv(DataReadyMessage) → relay.read(tensors)
```

Same mechanism for all data flows. Same-GPU stages use CUDA IPC (zero copy).

### 2.4 relay_io.py

Utility module providing:
- `write_payload` / `read_payload` — full StagePayload serialization via relay
- `write_blob` / `read_blob` — raw tensor transfer for streaming chunks
- `send_stream_chunk` — handles same-GPU IPC vs cross-GPU relay, NIXL credit deadlock avoidance
- `extract_tensors` / `restore_tensors` — recursive tensor extraction from nested dicts

## 3. Scheduling Layer

All schedulers share the same interface: `inbox`, `outbox`, `start()`, `stop()`, `abort()`.

### 3.1 OmniScheduler

For AR stages. Inherits from SGLang's Scheduler — reuses `get_next_batch_to_run()`, `run_batch()`, `process_batch_result()`, `event_loop_normal()`, overlap scheduling.

Overrides `__init__` (skip ZMQ/tokenizer/metrics), `recv_requests()` (drain inbox, route stream chunks to per-request state), `run_batch()` (delegate to model_runner).

Runs in dedicated thread. Stage communicates via thread-safe queues.

### 3.2 SimpleScheduler

For non-AR stages (preprocessing, encoders, aggregate, decode). No KV cache, no batching. Just `inbox.get() → fn(data) → outbox.put()`.

### 3.3 Code2WavScheduler

Streaming vocoder. Handles `new_request` → init, `stream_chunk` → accumulate + decode, `stream_done` → flush + output.

### 3.4 Message Types

```python
class IncomingMessage:
    request_id: str
    type: "new_request" | "stream_chunk" | "stream_done"
    data: Any

class OutgoingMessage:
    request_id: str
    type: "result" | "stream"
    data: Any
    target: str | None  # for stream: downstream stage name
```

### 4.1 ModelRunner (Base)

Shared execute pipeline for all AR models.

```python
class ModelRunner:
    def execute(self, scheduler_output):
        forward_batch = ForwardBatch.init_new(...)
        batch_result = self.prepare_forward(...)  # hook
        if batch_result is None:
            batch_result = self.tp_worker.forward_batch_generation(forward_batch)
        self.post_forward(batch_result, ...)       # hook
        # sample, logit processing, output extraction
        return ModelRunnerOutput(...)
```

Shared: ForwardBatch construction, sampling, repetition penalty, codec suppression, output processing.

### 4.2 ThinkerModelRunner

Injects multimodal embeddings (image/video/audio) + deepstack before forward.

```python
class ThinkerModelRunner(ModelRunner):
    def prepare_forward(self, ...):
        # Inject multimodal embeds into forward_batch
```

### 4.3 FeedbackARModelRunner

Shared model runner for all AR+codebook models (Qwen3 talker, Fish TTS, future models). The model's `forward()` handles backbone + secondary head internally. This runner writes/reads model buffers around forward.

```python
class FeedbackARModelRunner(ModelRunner):
    def __init__(self, tp_worker, output_processor, outbox, *,
                 write_buffers_fn, extract_output_fn, prefill_forward_fn=None):
        ...

    def prepare_forward(self, ...):
        if decode:
            self._write_buffers(model, schedule_batch, requests)
        elif prefill and self._prefill_forward:
            return self._prefill_forward(tp_worker, forward_batch, ...)
        return None

    def post_forward(self, ...):
        self._extract_output(model, schedule_batch, requests, outbox)
```

Model-specific behavior via three callbacks:
- `write_buffers_fn`: write previous step's feedback into model buffers
- `extract_output_fn`: read codes/feedback from model after forward
- `prefill_forward_fn`: custom forward for prefill (optional)

### 4.4 Callback Pattern

Each model provides a `callbacks.py` with three functions:

**Qwen3 Talker (`models/qwen3_omni/callbacks.py`):**
- `write_talker_buffers`: feedback_embeds + trailing/pad → `model._feedback_buffer`
- `extract_talker_output`: `model._output_codes` → outbox, `_output_embeds` → feedback
- `talker_prefill_forward`: projected input_embeds prefill

**Fish TTS (`models/fishaudio_s2_pro/callbacks.py`):**
- `write_fish_buffers`: codebook values → `model._vq_codes`
- `extract_fish_output`: `model._output_codes` → per-request output
- `fish_prefill_forward`: VQ embedding injection into input_embeds

Adding a third model = write a new `callbacks.py` with three functions.

### 4.5 Model: One Decode Step (AR + Codebook)

Both Qwen3 talker and Fish TTS follow the same internal pattern:

```
1. Read previous step's feedback from model buffers (written by FeedbackARModelRunner)
2. AR backbone → hidden states → logits
3. Sample first code from logits
4. Secondary head predicts remaining codebook layers autoregressively
5. Store combined output → buffers for next step
6. Output: multi-layer codes + feedback
```

The model class handles steps 1-6 inside `forward()`. FeedbackARModelRunner handles writing (before) and reading (after).

## 5. Model Directory Convention

Every model follows the same file structure:

```
models/<model_name>/
├── config.py              — Pipeline config (stage definitions, GPU placement)
├── stages.py              — Stage factories (returns callable or OmniScheduler)
├── routing.py             — Stage routing functions (which stage follows which)
├── request_builders.py    — Inter-stage data transform (build engine requests)
├── payload_types.py       — Model-specific pipeline state
├── callbacks.py           — FeedbackARModelRunner callbacks (write_buffers, extract_output)
├── __init__.py
└── components/            — Model-specific torch modules, preprocessors, encoders
```

### 5.1 Qwen3-Omni

```
models/qwen3_omni/
├── config.py              — 8-stage speech, 6-stage text
├── stages.py              — 8 factories
├── routing.py             — 8 routing functions
├── request_builders.py    — build thinker/talker/encoder requests
├── payload_types.py       — PipelineState, OmniEvent, ThinkerOutput
├── callbacks.py           — write_talker_buffers, extract_talker_output, talker_prefill_forward
├── hf_config.py           — HF config classes
├── merge.py               — Merge 3 encoder outputs for thinker
├── components/
│   ├── thinker.py         — Model loader/wrapper (Qwen3OmniSplitThinker)
│   ├── thinker_model.py   — SGLang thinker model definition
│   ├── talker.py          — SGLang talker model (fused MTP)
│   ├── preprocessor.py    — Tokenize, load media, apply HF processor
│   ├── image_encoder.py   — Image tower
│   ├── audio_encoder.py   — Audio tower
│   ├── talker_input.py    — Build talker prefill
│   ├── code2wav_scheduler.py — Vocoder streaming scheduler
│   └── common.py          — Shared helpers
```

**Speech pipeline (8 stages):**
```
preprocessing → image_encoder → audio_encoder → aggregate → thinker → decode → talker → code2wav
```

### 5.2 Fish Audio S2-Pro

```
models/fishaudio_s2_pro/
├── config.py              — 3-stage TTS
├── stages.py              — 3 factories
├── routing.py             — 3 routing functions
├── request_builders.py    — build_sglang_tts_request, apply_tts_result
├── payload_types.py       — S2ProState
├── callbacks.py           — write_fish_buffers, extract_fish_output, fish_prefill_forward
├── sglang_model.py        — SGLang model registration
├── tokenizer.py           — Tokenizer wrapper
└── fish_speech/           — Model definitions (text2semantic, DAC codec)
```

**Pipeline (3 stages):**
```
preprocessing → tts_engine → vocoder
```

### 6.1 PipelineConfig

Declarative. Each stage specifies factory function (string), routing function, relay config, stream targets. Factory returns a scheduler (has `inbox`/`outbox`) or a callable (wrapped in SimpleScheduler by compiler).

### 6.2 Compiler

`compile_pipeline(config)` → `(Coordinator, list[Stage])`. Duck-typed scheduler detection: `hasattr(result, "inbox") and hasattr(result, "outbox")`.

### 6.3 Multi-Process (mp_runner)

Each stage in its own OS process (`spawn` context). Main process runs Coordinator only.

## 7. Adding a New Model

1. Create `models/<name>/config.py` — define pipeline stages
2. Create `models/<name>/stages.py` — factory per stage (return callable or OmniScheduler)
3. Create `models/<name>/routing.py` — routing functions
4. Create `models/<name>/callbacks.py` — if AR+codebook: three functions for FeedbackARModelRunner
5. Create `models/<name>/components/` — model definitions, preprocessor
6. Register in model registry

Everything else (Stage, Coordinator, OmniScheduler, ModelRunner, relay, compiler) is reused as-is.
